src/kv/committable_tx.h

```diff
@@ -0,0 +1,363 @@
+            false);
+        }
+        catch (const std::exception& e)
+        {
+          committed = false;
+
+          LOG_FAIL_FMT("Error during serialisation");
+          LOG_DEBUG_FMT("Error during serialisation: {}", e.what());
+
+          // Discard original exception type, throw as now fatal
+          // KvSerialiserException
+          throw KvSerialiserException(e.what());
+        }
+      }
+    }
+
+    /** Get version at which this transaction was committed.
+     *
+     * Throws if this is not successfully committed - should only be called if
+     * an earlier call to commit() returned CommitResult::SUCCESS
+     *
+     * @return Commit version
+     */
+    Version commit_version()
+    {
+      if (!committed)
+        throw std::logic_error("Transaction not yet committed");
+
+      if (!success)
+        throw std::logic_error("Transaction aborted");
+
+      return version;
+    }
+
+    /** Get term in which this transaction was committed.
+     *
+     * Throws if this is not successfully committed - should only be called if
+     * an earlier call to commit() returned CommitResult::SUCCESS
+     *
+     * @return Commit term
+     */
+    Version commit_term()
+    {
+      if (!committed)
+        throw std::logic_error("Transaction not yet committed");
+
+      if (!success)
+        throw std::logic_error("Transaction aborted");
+
+      return view;
+    }
+
+    /** Version for the transaction set
+     *
+     * @return Committed version, or `kv::NoVersion` otherwise
+     */
+    Version get_version()
+    {
+      return version;
+    }
+
+    Version get_read_version()
+    {
+      return read_version.value_or(NoVersion);
+    }
+
+    Version get_max_conflict_version()
+    {
+      return max_conflict_version;
+    }
+
+    Version get_term()
+    {
+      return view;
+    }
+
+    void set_change_list(OrderedChanges&& change_list_, Term term_) override
+    {
+      // if all_changes is not empty then any coinciding keys will not be
+      // overwritten
+      all_changes.merge(change_list_);
+      view = term_;
+    }
+
+    void set_req_id(const kv::TxHistory::RequestID& req_id_)
+    {
+      req_id = req_id_;
+    }
+
+    const kv::TxHistory::RequestID& get_req_id()
+    {
+      return req_id;
+    }
+
+    void set_read_version_and_term(Version v, Term t)
+    {
+      if (!read_version.has_value())
+      {
+        read_version = v;
+        view = t;
+      }
+      else
+      {
+        throw std::logic_error("Read version already set");
+      }
+    }
+
+    void set_root_at_read_version(const crypto::Sha256Hash& r)
+    {
+      root_at_read_version = r;
+    }
+  };
+
+  // Used by frontend for reserved transactions. These are constructed with a
+  // pre-reserved Version, and _must succeed_ to fulfil this version. Otherwise
+  // they create a hole in the transaction order, and no future transactions can
+  // complete.
```

Comment:
```suggestion
  // complete. These transactions are used internally by CCF for the sole
  // purpose of recording node signatures and are safe in this particular
  // situation because they never perform any reads and therefore can
  // never conflict.
```